modules/programs/brew-bundle.nix

```diff
@@ -0,0 +1,225 @@
+# Created by: https://github.com/malob
+# Inspired by: https://github.com/lccambiaghi/nixpkgs/blob/main/modules/homebrew.nix
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.programs.brew-bundle;
+
+  brewfileSection = heading: type: entries:
+    if entries != [] then
+      "# ${heading}\n" + (concatMapStrings (name: "${type} \"${name}\"\n") entries) + "\n"
+    else "";
+
+  masBrewfileSection = entries:
+    if entries != {} then
+      "# Mac App Store apps\n" +
+      concatStringsSep "\n" (mapAttrsToList (name: id: ''mas "${name}", id: ${toString id}'') entries) +
+      "\n"
+    else "";
+
+  brewfile = pkgs.writeText "Brewfile" (
+    (brewfileSection "Taps" "tap" cfg.taps) +
+    (brewfileSection "Brews" "brew" cfg.brews) +
+    (brewfileSection "Casks" "cask" cfg.casks) +
+    (masBrewfileSection cfg.masApps) +
+    (brewfileSection "Docker contrainers" "whalebrew" cfg.whalebrews) +
+    (if cfg.extraConfig != "" then "# Extra config\n" + cfg.extraConfig else "")
+  );
+
+  brew-bundle-command =
+    "HOMEBREW_NO_AUTO_UPDATE=" +
+    (if cfg.noAutoUpdate then "1" else "0") +
+    " brew bundle --file='${brewfile}' --no-lock" +
+    (if cfg.cleanupType == "uninstall" || cfg.cleanupType == "zap" then " --cleanup" else "") +
+    (if cfg.cleanupType == "zap" then " --zap" else "");
+
+in
+
+{
+  options.programs.brew-bundle = {
+    enable = mkEnableOption ''
```

Comment:
I think having an enable option could potentially trip up user configs.  It would make any module that relies on this want to set enable = true, which might confuse the user if they have enable = false; alternately if the module doesn't set enable = true, the user might be confused why it isn't working.

I'd suggest the feature is disabled if no formulae/casks/etc. are specified.

Reply:
So I'm under the impression that having an `enable` option like this for a module if pretty idiomatic in the `Nix` ecosystem (though there are definitely exceptions), and so shouldn't trip up users. Also, a nice property of having an explicit enable option, is that it makes it easy to temporarily turn off the functionality the module provides without having to remove all the associated configuration of the various options.

Totally open to counter arguments, if you or others really feel like I'm off base here.

Reply:
I definitely agree that an enable option is pretty idiomatic with the rest of nix and should be included, even if true by default.

Reply:
As an added side effect it wouldn't be possible to "clear" brew formula since that would disable the brew steps altogether. I agree it's not the most intuitive pattern, but it is idiomatic and there's no great alternative given that home-brew would become a requirement otherwise.

It's always possible to add warnings/assertions for the opposite case, which doesn't make sense.

Reply:
Cool, let's stick with the Nix idiom here.